### PR TITLE
changed aliasing checks to properly recurse for computing last usage

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121692
* __->__ #122444
* #122474
* #121624

Fixes https://github.com/pytorch/pytorch/issues/122457

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by oulgen** [comment]
I think you should break this PR into two, one cosmetic changes, one functional changes.

**Comment by Chillee** [comment]
@oulgen done

**Review by yifuwang — APPROVED** [review]

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -7381,11 +7381,11 @@ def create(cls, collective_op: "TensorBox"):
 
     def get_inputs_that_alias_output(self):
         # Signal to codegen that our output buffer isn't safe to reuse
-        return [self.inputs[0].codegen_reference()]
+        return [self.inputs[0].get_name()]
```
Comment (yifuwang) [inline torch/_inductor/ir.py:7384]: This is deprecated code. But curious if you know if there was any significance of using `codegen_reference`.

Comment (Chillee) [reply torch/_inductor/ir.py:7384]: No. Pretty sure it never was right. See https://github.com/pytorch/pytorch/pull/108811#discussion_r1319338204

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**